arroyo/processing/strategies/streaming/transform.py

```diff
@@ -230,44 +230,83 @@ def parallel_transform_worker_apply(
     input_batch: MessageBatch[TPayload],
     output_block: SharedMemory,
     start_index: int = 0,
+    batch_execute: bool = False,
 ) -> Tuple[int, MessageBatch[TTransformed]]:
     output_batch: MessageBatch[TTransformed] = MessageBatch(output_block)
 
     i = start_index
-    while i < len(input_batch):
-        message = input_batch[i]
-
+    if batch_execute:
+        chunk = [input_batch[ind] for ind in range(len(input_batch) - start_index)]
         try:
-            result = function(message)
+            result = function(chunk)
         except Exception:
-            # The remote traceback thrown when retrieving the result from the
-            # pool elides a lot of useful data (and usually includes a
-            # truncated traceback), logging it here allows us to get this
-            # information at the expense of sending duplicate events to Sentry
-            # (one from the child and one from the parent.)
             logger.warning(
-                "Caught exception while applying %r to %r!",
+                "Caught exception while applying %r to %r messages!",
                 function,
-                message,
+                len(chunk),
                 exc_info=True,
             )
             raise
 
-        try:
-            output_batch.append(
-                Message(message.partition, message.offset, result, message.timestamp)
-            )
-        except ValueTooLarge:
-            # If the output batch cannot accept the transformed message when
-            # the batch is empty, we'll never be able to write it and should
-            # error instead of retrying. Otherwise, we need to return the
-            # values we've already accumulated and continue processing later.
-            if len(output_batch) == 0:
+        for ind, message in enumerate(chunk):
```

Comment:
This does not seem a clean way to accomplish your goal. 
By introducing a batching concept in the `parallel_transform_worker_apply` function we are making it twice as complex.
The same result can be achieved without touching the infrastructure at all by just passing batches of messages to the parallel transform step:
The message type for the parallel transform step is generic. You can pass whatever you want. 
What about batching messages together in a transform step before the parallel step, passing these batches to the parallel step, and eventually if you need, un-batch them in a step after the parallel step? 

This approach seems to have a few advantages:
- You maximize the gain from the parallel step. Big messages that take long to process parallelize better than smaller messages as the overhead of the parallelization counts less. So batching messages before sending them through the parallel step should help.
- This seems an ad hoc feature that does not need to be in the infrastructure as it is likely to be specific to the product feature you are building. We want to keep the infrastructure abstractions as simple as possible (which are already insanely too complex) so that they are easier to understand. In order for this to type check properly, the `function` passed to `parallel_transform_worker_apply` must be able to support two input types, `TPayload` and `Sequence[TPayload]`, which will have to be supported by all use cases that rely on arroyo.
- This implementation becomes more confusing as now it deals with two independent concepts of batch that mean very different things.